File: study/ts/script/scene/view/bout/PopupAttrOverView.ts
import { G_UserData } from "../../../init";
import CommonDesValue from "../../../ui/component/CommonDesValue";
import CommonNormalMiniPop4 from "../../../ui/component/CommonNormalMiniPop4";
import PopupBase from "../../../ui/PopupBase";
import { AttrDataHelper } from "../../../utils/data/AttrDataHelper";
import { table } from "../../../utils/table";
import { TextHelper } from "../../../utils/TextHelper";
import { BoutHelper } from "./BoutHelper";
const {ccclass, property} = cc._decorator;
@ccclass
export default class PopupAttrOverView extends PopupBase{
    name: 'PopupAttrOverView';
    @property({
        type: CommonNormalMiniPop4,
        visible: true
    })
    _commonNodeBk: CommonNormalMiniPop4 = null;
    @property({
        type: cc.Node,
        visible: true
    })
    _panelAttr: cc.Node = null;
    @property({
        type: cc.Label,
        visible: true
    })
    _boutName: cc.Label = null;

    @property({
        type: cc.Prefab,
        visible: true
    })
    _CommonDesValue: cc.Prefab = null;

    
    
    
    private _attrPanelSize:cc.Size;
    onCreate() {
        this._attrPanelSize = this._panelAttr.getContentSize();
        this._commonNodeBk.addCloseEventListener(()=> {
            this.close();
        });
    }
    onEnter() {
        this._updateUI();
        this._acquiredBout();
    }
    onExit() {
    }
    _acquiredBout() {
        var nameList = '';
        var boutList = G_UserData.getBout().getBoutList();
        var curBoutId = G_UserData.getBout().getCurBoutId();
        var boutInfo = G_UserData.getBout().getBoutInfo();
        var list = G_UserData.getBout().getBoutAcquire();
        var count = table.nums(list);
        if (boutList[curBoutId] && table.nums(boutList[curBoutId]) > 0) {
            count = count + 1;
        }
        count = cc.misc.clampf(1, count, table.maxn(boutInfo));
        for (var i = 1; i <= count; i++) {
            var info = BoutHelper.getBoutBaseItem(i);
            var ser = i == 1?(nameList):nameList + '、';
            if (i == 5) {
                ser = ser + '\n';
            }
            nameList = ser + info.name;
        }
        this._boutName.string = (nameList);
    }
    _updateUI() {
        this._panelAttr.removeAllChildren();
        var count = 1;
        var result = {};
        AttrDataHelper.appendAttr(result, G_UserData.getBout().getAttrSingleInfo());
        for (let k in result) {
            var v = result[k];
            var [attrName, attrValue] = TextHelper.getAttrBasicText(k, v);
            attrName = TextHelper.expandTextByLen(attrName, 4);
            var uiNode = cc.instantiate(this._CommonDesValue).getComponent(CommonDesValue);
            uiNode.updateUI(attrName + ': ', attrValue,null,4);
            var posX = count % 2 == 1 ? -143*1.2:143/2;
            var posY = - Math.ceil(count / 2) * 30;
            uiNode.node.setPosition(cc.v2(posX, posY));
            uiNode.setFontSize(20);
            this._panelAttr.addChild(uiNode.node);
            count = count + 1;
        }
    }
}

File: study/lua/src/app/scene/view/bout/PopupAttrOverView.lua
-- @Author  panhoa
-- @Date  4.3.2020
-- @Role 

local PopupBase = require("app.ui.PopupBase")
local PopupAttrOverView = class("PopupAttrOverView", PopupBase)
local CSHelper = require("yoka.utils.CSHelper")
local TextHelper = require("app.utils.TextHelper")
local AttrDataHelper = require("app.utils.data.AttrDataHelper")


function PopupAttrOverView:ctor()
    
    local resource = {
		file = Path.getCSB("PopupAttrOverView", "bout"),
	}
	PopupAttrOverView.super.ctor(self, resource, true, false)
end

function PopupAttrOverView:onCreate()
    self._attrPanelSize = self._panelAttr:getContentSize()
	self._commonNodeBk:addCloseEventListener(function( ... )
        -- body
        self:close()
    end)
end

function PopupAttrOverView:onEnter()
    self:_updateUI()
    self:_acquiredBout()
end

function PopupAttrOverView:onExit()
end

function PopupAttrOverView:_acquiredBout( ... )
    -- body
    local nameList = ""
    local BoutHelper = require("app.scene.view.bout.BoutHelper")
    local boutList = G_UserData:getBout():getBoutList()
    local curBoutId = G_UserData:getBout():getCurBoutId()
    local boutInfo = G_UserData:getBout():getBoutInfo()

    local list = G_UserData:getBout():getBoutAcquire()
    local count = table.nums(list)
    if boutList[curBoutId] and table.nums(boutList[curBoutId]) > 0 then
        count = (count + 1)
    end
    count = cc.clampf(1, count, table.maxn(boutInfo))

    for i=1, count do
        local info = BoutHelper.getBoutBaseItem(i)
        local ser = i == 1 and nameList or nameList.."、"
        if i == 5 then
            ser = ser.."\n"
        end
        nameList = ser..info.name
    end
    self._boutName:setString(nameList)
end

function PopupAttrOverView:_updateUI( ... )
    -- body
    self._panelAttr:removeAllChildren()
    local count = 1
    local result = {}
    AttrDataHelper.appendAttr(result, G_UserData:getBout():getAttrSingleInfo())
    
    for k,v in pairs(result) do
        local attrName, attrValue = TextHelper.getAttrBasicText(k, v)
        attrName = TextHelper.expandTextByLen(attrName, 4)
        local uiNode = CSHelper.loadResourceNode(Path.getCSB("CommonDesValue", "common"))
        uiNode:updateUI(attrName.."：", attrValue)
        local posX = (count % 2) == 1 and 90 or 345
        local posY = 225 - math.ceil(count / 2) * 30
        uiNode:setPosition(cc.p(posX, posY))
        uiNode:setFontSize(20)
        self._panelAttr:addChild(uiNode)
        count = (count + 1)
    end
end


return PopupAttrOverView